debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnection.java

```diff
@@ -473,6 +495,10 @@ public String buildSelectWithRowLimits(TableId tableId, int limit, String projec
         sql
                 .append(" ORDER BY ")
                 .append(orderBy);
+        if (this.optionRecompile) {
```

Comment:
@jpechane, IIUC, you had a slightly different solution in mind, right? In the issue you said:

> Extend the base class so the connectors can nrich the common generate SQL statement

I'd argue that's the case already; the methods `AbstractIncrementalSnapshotChangeEventSource#buildChunkQuery()` and `buildMaxPrimaryKeyQuery()` could be overridden in an implementation specific to SQL Server. Would you prefer that?

That said, what is "parameter sniffing" in this context, @Thiago-Dantas?

Reply:
@gunnarmorling Well, I think at the end this is fine. I forgot that it can be done at the datbase connection level.

Reply:
@gunnarmorling I'm just gonna defer the explanation to Brent Ozar @ https://www.brentozar.com/archive/2013/06/the-elephant-and-the-mouse-or-parameter-sniffing-in-sql-server/
"When the procedure is compiled or recompiled, the value passed into the parameter is evaluated and used to create an execution plan. That value is then stored with the execution plan in the plan cache. On subsequent executions, that same value – and same plan – is used.

This is a normal, expected behavior in SQL Server. Because compiling queries is expensive, you want plans stored in the cache. You want SQL Server to re-use them as much as possible.

But what happens when the values in a table you’re querying aren’t evenly distributed? What if one value would return 10 rows and another value would return 10,000 rows, or 10 million rows? I call this the elephant and the mouse problem. You would handle one animal differently than the other; SQL Server might create different plans for the queries. But it doesn’t, because you’re using parameters. "

The gist is if your table is not evenly distributed, which can be pretty common in N-N tables that have both IDs as PKs, it can end up with very poor execution plans

The JDBC transforms the query execution into a procedure(internal proc sp_executesql) call and is then susceptible to this

The query option OPTION(RECOMPILE) forces it to evaluate the parameters and build a new temporary, uncached, execution plan for the values passed.

I pasted warnings all over about CPU increase but in most cases that should not materialize as you would need to be running several queries per second with OPTION(RECOMPILE) to notice the increase in CPU

Reply:
Ok, thanks for clarifying, @Thiago-Dantas! Kinda suprising that the execution plan is determined only exactly once and never changed again.

@jpechane, so in that light, are you good with this change then? The test failure deserves logging in Jira though, we have seen it before, I don't think it's related.